src/open-url.coffee

```diff
@@ -15,40 +17,103 @@ bookmarks = {}
 				msg.send 'Oops?'
 			else
 				bookmarks = JSON.parse body
 
-	open = (key, callback) ->
-		if bookmarks[key]?
-			robot.http(bookmarks[key]).get() (err, res, body) ->
-				callback (not err?) and res.statusCode is 200
-		else
-			callback false
+	###*
+	* This creates a function to shim the http response
+	* @param {function} post successes to
+	* @param {function} post failures to
+	* @return {function} Takes (err, res, body) and passes them to (resolve, reject)
+	###
+	createResolver = (resolve, reject) ->
+		(err, res, body) ->
+			if (not err?) and (res.statusCode is 200)
+				resolve(body)
+			else
+				reject(err ? new Error("StatusCode: #{res.statusCode}"))
 
-	bookmark = (key, value, callback) ->
-		bookmarks[key] = value
-		robot.http("#{firebaseUrl}/data/.json?auth=#{firebaseAuth}")
-			.patch(JSON.stringify({ bookmarks: bookmarks })) (err, res, body) ->
-				callback (not err?) and res.statusCode is 200
+	###*
+	* This creates a function to report errors
+	* @param {Object} Hubot msg object
+	* @return {function} Takes (Error) and reports it
+	###
+	createErrorReporter = (context) ->
+		(error) ->
+			robot.logger.error(error)
+			context.send('Something went wrong. Debug output logged')
 
-	robot.respond /open (\S+)$/i, (context) ->
-		open context.match[1], (done) ->
-			context.send if done then context.random confirmations else "Couldn't find " + context.match[1] + ' key.'
+	###*
+	* Attempt to visit the url referenced by key
+	* @param {string} url to visit
+	* @return {Promise} Handle output from the asynchronous
+	###
+	open = (url) ->
+		new Promise (resolve, reject) ->
+			robot.http(url).get() createResolver(resolve, reject)
+
+	###*
+	* Store a url in the cache and firebase
+	* @param {string} key under which to store it
+	* @param {string} bookmark to store
+	* @return {Promise} Handle output from the asynchronous section
+	###
+	bookmark = (key, value) ->
+		new Promise (resolve, reject) ->
+			bookmarks[key] = value
+			robot.http("#{firebaseUrl}/data/.json?auth=#{firebaseAuth}")
+				.patch(JSON.stringify({ bookmarks: bookmarks })) createResolver(resolve, reject)
+
+	###*
+	* Extract a value from the context provided
+	* @param {Object} Hubot msg object
+	* @return {string} Value from the stored bookmarks
+	###
+	getValueFromContext = (context) ->
+		if bookmarks[context.match[1]]?
+			return bookmarks[context.match[1]]
+		else if bookmarks[context.envelope.room]?
+			return bookmarks[context.envelope.room]
+		else
+			throw new Error("Couldn't find key.")
 
-	robot.respond /open$/i, (context) ->
-		open context.envelope.room, (done) ->
-			context.send if done then context.random confirmations else "Couldn't find " + context.envelope.room + ' key.'
+	###*
+	* Attempt to open the specified bookmark, defaulting to a bookmark for the room
+	* (?:\W(\w+))? match up to the first word after open, capturing just the word
+	###
+	robot.respond /open(?:\W(\w+))?/i, (context) ->
+		context.send(context.random(holdings))
```

Comment:
Nope, nope, nope.
That is beyond the scope of current Adapters.  But would be very cool.  In fact, if you want to Hack Friday the flowdock adapter the module is here https://github.com/resin-io-modules/hubot-flowdock